### DISCUSSION TITLE
How do we get ort to support a custom ONNX operator?

### DISCUSSION THREAD

**Discussion by ksngc** [pykeio/ort#151] [Q&A] [answered]
The page here: https://github.com/onnx/tutorials/tree/main/PyTorchCustomOperator outlines how to:

(1) design a new PyTorch operator
(2) how to write a custom ONNX operator corresponding to this PyTorch operator + export this to a model
(3) how to write this same ONNX operator in ONNX Runtime via C++ and then run inference

What's the easiest way to do (3), but using this library (i.e., ort for Rust)?

**Comment by decahedron1** [comment, marked as answer]
> (3) how to write this same ONNX operator in ONNX Runtime via C++ and then run inference
>
> What's the easiest way to do (3), but using this library (i.e., ort for Rust)?

You can achieve this by writing a custom operator library with the C++ API. See [onnxruntime-extensions](https://github.com/microsoft/onnxruntime-extensions) for an example of how this is done.

To then load the custom operator library in `ort`, enable `ort`'s `custom-ops` feature:

```toml
[dependencies]
ort = { version = "2.0.0-rc.0", features = [ "custom-ops" ] } 
```

In your code, register your custom operator's library on the `SessionBuilder`:
```rs
let session = Session::builder()?
    .with_custom_ops_lib("my_ops_lib")?
    .with_model_from_file("example.onnx")?;
```
